pkg/metrics/builder/builder_test.go

```diff
@@ -0,0 +1,226 @@
+package builder
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+)
+
+func getDescStruct(opts MetricOpts) string {
+	var labels []string
+	lpStrings := make([]string, 0, len(opts.labelKeyValues))
+	for key, value := range opts.labelKeyValues {
+		lpStrings = append(lpStrings, fmt.Sprintf("%s=%q", key, value))
+	}
+	sort.Strings(lpStrings)
+
+	return fmt.Sprintf("Desc{fqName: %q, help: %q, constLabels: {%s}, variableLabels: %v}",
+		opts.name, opts.desc, strings.Join(lpStrings, ","), labels)
+}
+
+func getCollectorDescription(collector prometheus.Collector) string {
+	switch reflect.TypeOf(collector).Elem().Name() {
+	case "histogram":
+		return collector.(prometheus.Histogram).Desc().String()
+	case "counter":
+		return collector.(prometheus.Counter).Desc().String()
+	default:
+		return ""
+	}
+}
+
+func TestMetricBuilder(t *testing.T) {
+	cases := []struct {
+		name                       string
+		collectorType              string
+		opts                       MetricOpts
+		expectedBuilderError       bool
+		expectedPromCollectorError bool
+		expectedErrorMessage       string
+	}{
+		{
+			name:          "Test histogram creation",
+			collectorType: "histogram",
+			opts: MetricOpts{
+				labels:         []string{"test1", "test2"},
+				labelKeyValues: map[string]string{"test1": "test1", "test2": "test2"},
+				desc:           "test histogram metrics",
+				name:           "test_histogram",
+				buckets:        []float64{10, 20, 30},
+				metricType:     Histogram,
+			},
```

Comment:
No user outside this package will be able to initialize the metric opts, since all the fields are package private.

Reply:
Not sure if we should allow the installer to create metrics or define all the metrics possible in the package and have the installer update the label values. I would assume we need to restrict the number of metrics that the installer pushes.